### ISSUE TITLE
What if the loss functions actually returned the loss?

### ISSUE THREAD

**Issue by oysteijo** [oysteijo/simd_neuralnet#3]
The loss function of a neural network is declared with the following prototype:
`void loss_func(const unsigned int n, const float *y_pred, const float *y_real, float *loss );`

would it not be cool and maybe even natural if this function actually returned the loss? (That's a scalar value?)

`float loss_func(const unsigned int n, const float *y_pred, const float *y_real, float *loss );`

**Comment by oysteijo** [comment]
The downside of this suggestion is that this calculation will take some cycles and the result will just be discarded in the backpropagation anyway. Maybe it is better to make a reporting function as needed?

The reporting function can be looked up based on the loss_func pointer.

    printf("loss: %f\n", get_loss_calculation( nn->loss )( n, y_pred, y_real ));

**Comment by oysteijo** [comment]
No.... I'm rethinking this. That is the job of the metric functions. I've now implemented the **metric** functions that supports the implemented loss functions. Closing issue.
